File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/PartitionLabels.java
package com.only4.algorithm.version4j.leetcode;

import java.util.ArrayList;
import java.util.List;

public class PartitionLabels {
    public List<Integer> partitionLabels(String s) {
        if (s.isEmpty()) return new ArrayList<>();

        int[] lastPositions = new int[26];

        for (int i = 0; i < s.length(); i++) {
            lastPositions[s.charAt(i) - 'a'] = i;
        }

        List<Integer> result = new ArrayList<>();
        int start = 0;
        int end = 0;

        for (int i = 0; i < s.length(); i++) {
            end = Math.max(end, lastPositions[s.charAt(i) - 'a']);

            if (i == end) {
                result.add(end - start + 1);
                start = i + 1;
            }
        }
        return result;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/partitionLabels.kt
package com.only4.algorithm.version4k.leetcode

fun partitionLabels(s: String): List<Int> {
    if (s.isEmpty()) return emptyList()

    val lastPositions = IntArray(26)

    for (i in s.indices) {
        lastPositions[s[i] - 'a'] = i
    }

    val result = mutableListOf<Int>()
    var start = 0
    var end = 0

    for (i in s.indices) {
        end = maxOf(end, lastPositions[s[i] - 'a'])

        if (i == end) {
            result.add(end - start + 1)
            start = i + 1
        }
    }
    return result
}
